ftw/simplelayout/browser/resources/RequestTracker.js

```diff
@@ -0,0 +1,107 @@
+(function(global, $) {
+
+  "use strict";
+
+  var URLWhiteList = [
+    "sl-ajax-reload-block-view",
+    "sl-toolbox-view",
+    "sl-ajax-delete-blocks-view",
+    "add_block",
+    "sl-ajax-edit-block-view",
+    "sl-ajax-upload-block-view",
+    "edit.json",
+    "tiny_mce_gzip.js"
+  ];
+
+  var URLBlackList = [
+    "@@z3cform_validate_field"
+  ];
+
+  function getParameterByName(name, url) {
+    name = name.replace(/[\[]/, "\\[").replace(/[\]]/, "\\]");
+    var regex = new RegExp("[\\?&]" + name + "=([^&#]*)");
+    var results = regex.exec(url);
+    return results === null ? "" : decodeURIComponent(results[1].replace(/\+/g, " "));
+  }
+
+  function Request(url, callback, options) {
+
+    options = $.extend({ threshold: 300 }, options);
+
+    var threshold = getParameterByName("threshold", url) || options.threshold;
+    var timer;
+
+    function isThrottled() {
+      var spinnerFlag = getParameterByName("spinner", url);
+      return spinnerFlag === "" && spinnerFlag !== "false";
+    }
+
+    function throttle() {
+      if(isThrottled()) {
+        timer = global.setTimeout(callback, threshold);
+      }
+    }
+
+    function cancel() { global.clearTimeout(timer); }
+
+    throttle();
+
+    return Object.freeze({
+      throttle: throttle,
+      cancel: cancel
+    });
+
+  }
+
+  function RequestTracker() {
+
+    var markerClass = "ajax-loading";
+    var $document = $(document);
+    var root = $(":root");
+
+    var requests = {};
+
+    function checkURL(url) {
+      var isWhiteListed = !!$.grep(URLWhiteList, function(whiteListEntry) {
+        return url.indexOf(whiteListEntry) >= 0;
+      }).length;
+
+      var isBlackListed = !!$.grep(URLBlackList, function(blackListEntry) {
+        return url.indexOf(blackListEntry) >= 0;
+      }).length;
+
+      return isWhiteListed && !isBlackListed;
+    }
+
+    function mark() { root.addClass(markerClass); }
+
+    function unmark() { root.removeClass(markerClass); }
+
+    function track(url) {
+      if(checkURL(url)) {
+        requests[url] = Request(url, mark);
+      }
+    }
+
+    function untrack(url) {
+      if(requests[url]) {
+        requests[url].cancel();
+        delete requests[url];
+        unmark();
```

Comment:
When having concurrent requests, the first finished request will unmark and the spinner will disappear, even when more tracked requests are pending.
This is an "academical" issue, since the user cannot click when the first request begins and thus the user can not trigger more requests.
But depending on the implementation, more requests may be triggered automatically.
If we can fix it easily and without refactoring everything, I would like to do it.

Proposal:
```diff
     function untrack(url) {
       if(requests[url]) {
         requests[url].cancel();
         delete requests[url];
-        unmark();
       }
+      if (Object.keys(requests).length === 0) {
+        unmark();
+      }
     }
```
The idea is that we only remove the spinner / unmark when all tracked requests are finished.
@bierik What do you think?
⚠️ `Object.keys(requests).length` may not be the best solution; I'm no JavaScript expert.